src/tunnel.cc

```diff
@@ -1006,6 +1007,18 @@ tunnelErrorComplete(int fd/*const Comm::ConnectionPointer &*/, void *data, size_
         tunnelState->server.conn->close();
 }
 
+bool
+TunnelStateData::checkRetry()
+{
+    if (shutting_down)
+        return false;
+    if (!FwdState::EnoughTimeToReForward(startTime))
+        return false;
+    if (noConnections())
```

Comment:
retry() calls closeServerConnection() _after_ checkRetry(). Would that make this check ineffective in some cases?

* Should we only check the client connection here instead?
* Or perhaps closeServerConnection() unconditionally, before calling checkRetry()? Or is that too risky for retry() callers? IIRC, we use (the equivalent of) this approach in master.